component/src/main/java/org/wso2/carbon/identity/authenticator/office365/Office365Authenticator.java

```diff
@@ -0,0 +1,339 @@
+        return Office365AuthenticatorConstants.AUTHENTICATOR_FRIENDLY_NAME;
+    }
+
+    /**
+     * Get the name of the Authenticator
+     */
+    @Override
+    public String getName() {
+        return Office365AuthenticatorConstants.AUTHENTICATOR_NAME;
+    }
+
+    /**
+     * Get Configuration Properties
+     */
+    @Override
+    public List<Property> getConfigurationProperties() {
+
+        List<Property> configProperties = new ArrayList<>();
+
+        Property clientId = new Property();
+        clientId.setName(OIDCAuthenticatorConstants.CLIENT_ID);
+        clientId.setDisplayName("Client Id");
+        clientId.setRequired(true);
+        clientId.setDescription("Enter Office365 client identifier value");
+        clientId.setDisplayOrder(0);
+        configProperties.add(clientId);
+
+        Property clientSecret = new Property();
+        clientSecret.setName(OIDCAuthenticatorConstants.CLIENT_SECRET);
+        clientSecret.setDisplayName("Client Secret");
+        clientSecret.setRequired(true);
+        clientSecret.setConfidential(true);
+        clientSecret.setDescription("Enter Office365 client secret value");
+        clientSecret.setDisplayOrder(1);
+        configProperties.add(clientSecret);
+
+        Property redirectUri = new Property();
+        redirectUri.setDisplayName("Callback Url");
+        redirectUri.setName(IdentityApplicationConstants.OAuth2.CALLBACK_URL);
+        clientId.setRequired(true);
+        redirectUri.setDescription("Enter value corresponding to callback url.");
+        redirectUri.setDisplayOrder(2);
+        configProperties.add(redirectUri);
+
+        return configProperties;
+    }
+
+    /**
+     * Get the Context identifier sent with the request.
+     */
+    @Override
+    public String getContextIdentifier(HttpServletRequest request) {
+        return request.getParameter(Office365AuthenticatorConstants.STATE);
+    }
+
+    /**
+     * Check whether the authentication or logout request can be handled by the authenticator
+     */
+    @Override
+    public boolean canHandle(HttpServletRequest arg0) {
+        return true;
+    }
+
+    /**
+     * Authenticator flow process
+     */
+    @Override
+    public AuthenticatorFlowStatus process(HttpServletRequest request, HttpServletResponse response,
+                                           AuthenticationContext context)
+            throws AuthenticationFailedException, LogoutFailedException {
+        // if the logout request comes, then no need to go through and doing complete the flow.
+        if (context.isLogoutRequest()) {
+            return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
+        }
+        if (StringUtils.isNotEmpty(request.getParameter(Office365AuthenticatorConstants.CODE))) {
+            processAuthenticationResponse(request, response, context);
+            return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
+        } else {
+            initiateAuthenticationRequest(request, response, context);
```

Comment:
Else condition is not required here since if it is true then it will return. Can have this below the else.

Reply:
In this case, It calls process method first and checks it has "code" in the request, very first time, request does not have "code", so we need to call 'initiateAuthenticationRequest' method to get "code", once gets the "code", it will call "processAuthenticationResponse" method to get access token and user details.

Reply:
Yes. But if it enters the true condition, then it will return and will not execute the below. But for else it will enter the "else" condition and call the initiateAuthenticationRequest method and then proceed. But the return condition is outside. Therefore we do not need to else condition. IMO we can have the initiateAuthenticationRequest after the if condition.

Reply:
Yes, you are correct and find the fixed in https://github.com/wso2-extensions/identity-outbound-auth-office365/pull/3